### PR TITLE
feat: separate of concern

### PR DESCRIPTION
**Pull request by magiskboy**

### CODE REVIEW

File: app/models/matching.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+from typing import Type
+import asyncio
+from dataclasses import dataclass
+from ._base import Computation
+from ..distances import DistanceCalculator
+from ..embeddings import Embedding
+from ..llm import LLAMA2Service
+from ..utils import split_sentences
+
+@dataclass()
```
Comment (marcusschiesser) [inline app/models/matching.py]: why not using pydantic?

File: app/utils.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+def split_sentences(document: str) -> list[str]:
+    return document.split('.')
```
Comment (marcusschiesser) [inline app/utils.py]: probably good to `trim` each sentence

File: app/models/_base.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+class Computation:
```
Comment (marcusschiesser) [inline app/models/_base.py]: 1. why not using https://andrewbrookins.com/technology/building-implicit-interfaces-in-python-with-protocol-classes/ ?
2. add typings

also, `compute` is the wrong abstraction. The types are different in `chat.py` and `matching.py`

File: app/models/chat.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from typing import AsyncIterator
+from ._base import Computation
+from .matching import ScoredSentences
+from ..llm import LLMService
+
+class ChatModel(Computation):
+    def __init__(self, llm: LLMService):
+        self.llm = llm
+
+    async def compute(self, sentences: list[ScoredSentences], query: str) -> AsyncIterator[str]:
```
Comment (marcusschiesser) [inline app/models/chat.py]: scores are not needed here, rename `sentences` to `context`

Comment (marcusschiesser) [reply app/models/chat.py]: to do RAG, we should actually call the LLM with a prompt that contains the sentences we retrieved before as context

```
join_context = ".\n".join(sentences)
prompt = f"""
    You are giving the context and the question, try to generate the answer with the given context:
    context: {join_context}.
    question: {query}."""
```

Comment (magiskboy) [reply app/models/chat.py]: It's so cool, thanks for this comment 💯

**Pull request merged**